hypothesis-python/src/hypothesis/searchstrategy/misc.py

```diff
@@ -85,3 +86,69 @@ def calc_is_cacheable(self, recur):
 
     def do_draw(self, data):
         return d.choice(data, self.elements)
+
+    def do_filtered_draw(self, data, filter):
+        # Set of indices that have been tried so far, so that we never test
+        # the same element twice during a draw.
+        known_bad_indices = set()
+
+        def check_index(i):
+            """Return ``True`` if the element at ``i`` satisfies the filter
+            condition.
+            """
+            if i in known_bad_indices:
+                return False
+            ok = filter.condition(self.elements[i])
+            if not ok:
+                if not known_bad_indices:
+                    filter.note_retried(data)
+                known_bad_indices.add(i)
+            return ok
+
+        # Start with ordinary rejection sampling. It's fast if it works, and
+        # if it doesn't work then it was only a small amount of overhead.
+        for _ in hrange(3):
```

Comment:
It seems that either this should use the `result = _.do_filtered_draw(...); if result is not self.not_satisfied:` idiom, or that idiom isn't particularly reusable.

I'm not sure which, but it feels like `default_do_filtered_draw` only exists because there's some confusion between `self` and `filter` as to which is actually the main strategy here.

Reply:
The reason for effectively reimplementing `default_do_filtered_draw` inline here is so that we can do things *slightly* differently.

Note that we use `check_index` instead of calling `filter.condition` directly, so that the filter outcome is saved in `known_bad_indices`. That wouldn't be possible if we delegated to `default_do_filtered_draw` as-is.

(And even if we had a custom `condition` hook on `default_do_filtered_draw`, it would have to be expecting drawn values, whereas we need raw indices here.)